tests/test_functional_api.py

```diff
@@ -0,0 +1,1251 @@
+        if "limit" not in extra.keys() and "name" not in extra.keys():
+            # History calls that use 'name' don't need the limit parameter
+            extra["limit"] = self.history_size * 2
+
+        # Fake history entries never have files, but randomize del_files anyway
+        if "name" in extra.keys() and "del_files" not in extra.keys():
+            if extra["name"] == "delete":
+                extra["del_files"] = randint(0, 1)
+        return self._get_api_json("history", extra_args=extra)
+
+    def _create_random_queue(self, minimum_size):
+        """
+        Ensure the queue has a minimum number of jobs entries, adding random new
+        jobs as necessary; excess jobs are not trimmed. Note that while the
+        queue is paused overall to prevent downloading, the individual jobs do
+        not have their priority set to paused.
+        """
+        # Make sure the queue is paused
+        assert self._get_api_json("pause")["status"] is True
+
+        # Only add jobs if we have to, generating new ones is expensive
+        queue_size = len(self._get_api_json("queue")["queue"]["slots"])
+        if queue_size >= minimum_size:
+            return
+        else:
+            minimum_size -= queue_size
+
+        charset = ascii_lowercase + digits
+        for _ in range(0, minimum_size):
+            job_name = "%s-CRQ" % ("".join(choice(charset) for i in range(16)))
+            job_dir = os.path.join(self.daemon_basedir, job_name)
+
+            # Create the job_dir and fill it with a bunch of smallish files with
+            # random names, sizes and content. Note that some of the tests
+            # expect at least two files per NZB.
+            try:
+                os.mkdir(job_dir)
+                for number_of_files in range(0, randint(2, 4)):
+                    job_file = "%s.%s" % ("".join(choice(charset) for i in range(randint(6, 18))), sample(charset, 3))
+                    with open(os.path.join(job_dir, job_file), "wb") as f:
+                        f.write(os.urandom(randint(1, 512 * 1024)))
+            except Exception:
+                pytest.fail("Failed to create random queue stuffings")
+
+            # Fabricate the NZB
+            nzb_file = create_nzb(job_dir)
+
+            # Add job to queue
+            assert (
+                self._get_api_json("addlocalfile", extra_args={"name": nzb_file, "nzbname": job_name})["status"] is True
+            )
+
+            # Remove cruft
+            try:
+                shutil.rmtree(job_dir)
+            except Exception:
+                warn("Failed to remove %s" % job_dir)
+
+    def _purge_queue(self, del_files=0):
+        """ Clear the entire queue """
+        self._get_api_json("queue", extra_args={"name": "purge", "del_files": del_files})
+        assert len(self._get_api_json("queue")["queue"]["slots"]) is int(0)
+
+
+@pytest.fixture(scope="class")
+def api_server(request):
```

Comment:
Hmmm, seems similar to the code in `conftest.py`. Nothing we can reuse to generalize this?

Reply:
I refrained from modifying start_sabnzbd_and_selenium() because I'd end up introducing bugs in the existing functional tests while making my own life harder. Now that the api tests are mostly complete and given the amount of overlap, the api_server() code should be integrated in conftest.py though. I'll get to work on that.

Reply:
Scope differences don't help with this. I wrote api_server() using a class scope so the server config can easily be tuned from the test class and a fake history db added at startup. As the api tests require just sab itself, there wasn't much of a penalty.

For start_sabnzbd_and_selenium() on the other hand, scope is set to session, probably in order to avoid having to start sab, chrome, and sabnews multiple times during a pytest run. Mixing scopes (or dynamically setting them) gets ugly real quick, and moving everything to scope class isn't exactly without penalty, so I'll need to change api_server() to use session scope and make it trigger stop+start of the sab instance whenever a "refresh" of the fake history db is needed.